### ISSUE TITLE
LlamaTpPartModel如何使用

### ISSUE THREAD

**Issue by feifeibear** [ModelTC/lightllm#287]
我照着llama2的单测，写了一个example测试`LlamaTpPartModel`生成结果。我发现它的输出没有什么意义。

这是huggingface generate输出
hf model output ['', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', 'He', 'is', 'working', 'on', '', '2', '0', '1', '9', '.', '\n', 'I', "'", 'm', 'not', 'sure', 'what', 'you', 'mean', 'by', '"', 'working', 'on', '', '2', '0', '1', '9', '".', '\n', 'I', "'", 'm', 'not', 'sure', 'what', 'you']

这是LlamaTpPartModel的输出
**['....... and and and and and and and and and and and and and the.\n\n\n\n...\n\n\n\n']**

下面测试脚本可以复现，LLAMA_MODEL_DIR改成本地llama2模型地址。

import numpy as np
from multiprocessing import Queue
import multiprocessing

from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer

LLAMA_MODEL_DIR="/share_nfs/fangjiarui/root/code/hf_models/llama-2-7b-hf/"

def test_model_inference(world_size, model_dir, model_class, batch_size, input_len, output_len, mode):
    ans_queue = Queue()
    workers = []
    for rank_id in range(world_size):
        model_kvargs = {
            "tp_rank": rank_id,
            "world_size": world_size,
            "weight_dir": model_dir,
            "max_total_token_num":batch_size * (input_len + output_len),
            "load_way": "HF",
            "mode": mode,
            "max_req_num": batch_size,
            "max_seq_length": (input_len + output_len)
        }
        
        proc = multiprocessing.Process(target=tppart_model_infer, args=(model_class, model_kvargs, batch_size, input_len, output_len, ans_queue))
        proc.start()
        workers.append(proc)

    for proc in workers:
        proc.join()

    assert not ans_queue.empty()
    while not ans_queue.empty():
        assert ans_queue.get()
    return 

def tppart_model_infer(model_class, model_kvargs, batch_size, input_len, output_len, ans_queue):
    import torch
    import torch.distributed as dist
    rank_id = model_kvargs["tp_rank"]
    world_size = model_kvargs["world_size"]

    dist.init_process_group('nccl', init_method='tcp://127.0.0.1:28765', rank=rank_id, world_size=world_size)
    torch.cuda.set_device(rank_id)

    import torch.distributed as dist
    dist.barrier()
    torch.cuda.empty_cache()

    model_part = model_class(model_kvargs)
    # warm up
    # test_data = np.vstack([np.arange(5, input_len + 5) for _ in range(batch_size)])
    # test_data = test_data.reshape(-1)
    # test_data = torch.from_numpy(test_data).cuda()
    # print(f"test_data shape {test_data.shape}")
    
    input_sentences = [
        "He is working on " for _ in range(batch_size)
    ]
    tokenizer = AutoTokenizer.from_pretrained(model_kvargs["weight_dir"])
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    
    tokenizer.padding_side = "left"
    
    input_tokens = tokenizer.batch_encode_plus(
                    input_sentences,
                    return_tensors="pt",
                    padding="max_length",
                    max_length=input_len,
                    truncation=True,
                )
    input_tokens["input_ids"] = input_tokens["input_ids"].to("cuda")
    input_tokens["attention_mask"]=input_tokens["attention_mask"].to("cuda")
    print(f'input_tokens shape {input_tokens["input_ids"].shape}')
    test_data = input_tokens["input_ids"].cuda().reshape(-1)
    
    # a reference model
    USE_REF = True
    if USE_REF:
        hf_model = AutoModelForCausalLM.from_pretrained(model_kvargs["weight_dir"], torch_dtype=torch.bfloat16)
        hf_model = hf_model.eval().to("cuda")
        output = hf_model.generate(**input_tokens, max_new_tokens=output_len, 
                                return_dict_in_generate=True, output_scores=True)
        output = output.sequences[0]
        output = tokenizer.batch_decode(output, skip_special_tokens=True)
        print("hf model output {}".format(output))

    
    b_req_idx = model_part.req_manager.alloc(batch_size).int()
    b_start_loc = torch.zeros(batch_size, dtype=torch.int32, device="cuda")
    b_seq_len = torch.zeros(batch_size, dtype=torch.int32, device="cuda")
    for i in range(batch_size):
        b_start_loc[i] = i * input_len
        b_seq_len[i] = input_len

    predict_res = []
    total_token_num = input_len * batch_size
    logics = model_part.forward(batch_size, 
                                total_token_num, 
                                input_len, 
                                test_data,
                                b_req_idx,
                                b_start_loc,
                                b_seq_len,
                                is_prefill=True)
    prob_out = torch.softmax(logics, dim=-1)
    predict_ids = torch.argmax(prob_out, dim=1, keepdim=True)
    predict_ids = predict_ids.detach().cpu().numpy()
    print(f"predict_ids shape after prefill {predict_ids.shape}")
    predict_res = predict_ids
    
    for i in range(output_len):
        b_start_loc = b_start_loc + torch.arange(0, batch_size, dtype=torch.int32, device="cuda")
        total_token_num += batch_size
        b_seq_len += 1
        logics = model_part.forward(batch_size, 
                                    total_token_num, 
                                    max_len_in_batch = input_len + i + 1, 
                                    input_ids = torch.from_numpy(predict_ids).cuda().reshape(-1), 
                                    b_req_idx = b_req_idx, 
                                    b_start_loc = b_start_loc, 
                                    b_seq_len = b_seq_len, 
                                    is_prefill=False)
        prob_out = torch.softmax(logics, dim=-1)
        predict_ids = torch.argmax(prob_out, dim=1, keepdim=True)
        predict_ids = predict_ids.detach().cpu().numpy()
        # print(f"predict_ids shape in decode step {predict_ids.shape}")
        predict_res = np.concatenate((predict_res, predict_ids), axis=-1)
    
    print(f"predict_res {predict_res}")
    predict_resp = tokenizer.batch_decode(predict_res, skip_special_tokens=True)
    print(f"predict_resp {predict_resp}")
    
    model_part.mem_manager.free_all()
    model_part.req_manager.free_all()
    dist.barrier()
    torch.cuda.synchronize()

import os
import sys
import unittest

class TestLlama2Infer(unittest.TestCase):

    def test_llama2_infer(self):
        from lightllm.models.llama.model import LlamaTpPartModel
        test_model_inference(world_size=1, 
                             model_dir=LLAMA_MODEL_DIR, 
                             model_class=LlamaTpPartModel, 
                             batch_size=1, 
                             input_len=32, 
                             output_len=32,
                             mode=[])
        return

if __name__ == '__main__':
    unittest.main()

**Comment by hiworldwzj** [comment]
@feifeibear 是不是你的这种写法有pad，还有 huggingface 的推理 还有 attention mask 的介入。能不能搞个没有pad的input_ids

**Comment by feifeibear** [comment]
> @feifeibear 是不是你的这种写法有pad，还有 huggingface 的推理 还有 attention mask 的介入。能不能搞个没有pad的input_ids

huggingface的用法没问题，我按lightllm的test写的输出有问题。你能给一个LlamaTpPartModel  example么？

**Comment by hiworldwzj** [comment]
@feifeibear 我的意思是huggingface 的那种实现，可能在计算 attention的时候，有mask 操作，因为lightllm 内部是nopad的，所以你的input_ids 最好没有 pad token。 这样应该就可以得到一个比较正常的输出了。

**Comment by feifeibear** [comment]
> @feifeibear 我的意思是huggingface 的那种实现，可能在计算 attention的时候，有mask 操作，因为lightllm 内部是nopad的，所以你的input_ids 最好没有 pad token。 这样应该就可以得到一个比较正常的输出了。

理解了！我改一下
